sycl/include/CL/sycl/detail/pi.hpp

```diff
@@ -165,24 +173,29 @@ namespace RT = cl::sycl::detail::pi;
 #define PI_TRACE(func) RT::Trace<decltype(func)>(func, #func)
 
 // This does the call, the trace and the check for no errors.
-#define PI_CALL(pi)                                                            \
-  RT::initialize(), RT::PiCall(#pi).check<cl::sycl::runtime_error>(            \
-                        RT::cast<detail::RT::PiResult>(pi))
+// Need to call initialize so that the ptrs are updated and then stored.
+#define PI_CALL(pi, ...)                                                       \
```

Comment:
I think PI_CALL(api)(args) is a more clear syntax than PI_CALL(api, args), can we continue to use that?
Actually now that you don't need the pre-processor "#pi", can we turn this into a C++ class and have it look like PiCall<api>(args) ? (e.g. similar to how the pi::Trace is written, and maybe even just use that class directly)

Reply:
Currently the api name is passed, and we are converting it into the variable name in the struct.
The token pasting operator # is not explicitly used, but it is being done. 
Eg:
PI_CALL(api, args) PI_TRACE<...>(PlunginInterface.PiFunctionTable.api) .
The macro is easy to call. Plus it implements checking of the result in the same call.

To use a class call, we will have to make the user pass the pointer to the calling API.
As: piCall(PluginInterface.PiFunctionTable.api)(arguments);

With a class call, we will need to change these call semantics and implement another function to check the result returned.
Is this acceptable? If yes, then we can do that.

NOTE: In the future, the user will need to pass the platform object to the macro, to get access to the PluginInterface object associated with a platform.
PI_CALL(platform, api, arguments);

The class call semantics will change to:
piCall(platform.PluginInterface.PiFunctionTable.api)(arguments);

Reply:
> To use a class call, we will have to make the user pass the pointer to the calling API.
As: piCall(PluginInterface.PiFunctionTable.api)(arguments);

I think you can use template specialization to map a PI api to its pointer in the PI functions table.

> With a class call, we will need to change these call semantics and implement another function to check the result returned.

It should be possible to run the checks in destructor of an automatic object handling the call/trace, or we could indeed be explicit and use piCall<api>(args).check()

Reply:
Sorry, I was a little ahead in my mind. 
I thought I uploaded my next patch, and started to talk about the PluginInformation structure I created in it.

> > To use a class call, we will have to make the user pass the pointer to the calling API.
> > As: piCall(PluginInterface.PiFunctionTable.api)(arguments);
> 
> I think you can use template specialization to map a PI api to its pointer in the PI functions table.
> 
Yes. I can do that. I will upload a new patch for this change.
> > With a class call, we will need to change these call semantics and implement another function to check the result returned.
> 
> It should be possible to run the checks in destructor of an automatic object handling the call/trace, or we could indeed be explicit and use piCall(args).check()

Yes. Makes sense.

Reply:
@smaslov-intel : I will be making a new patch for this change. There is another patch on the way , that builds up on this change. 
The implementation of the class may change further after the next patch.
Let me know if you are Ok with it.